### PR TITLE
Added initial doc about engine code structure

### PR DESCRIPTION
**Pull request by JCash**
## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/docs/CODE_STRUCTURE.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+E.g. `render` may use `graphics`, but the reverse order is not allowed.
+
+* testmain - Helper library for unit tests and platform specific functionality.
+* dlib - Our main util library with containers, sockets, I/O, compression etc
+* texc - Our texture compiler. Creates a .jar+.dll for use with bob.jar
+* [modelc](../modelc/README.md) - Our model compiler. Creates a .jar+.dll for use with bob.jar
+* ddf - Our runtime data format. It's a slimmed down version of Protobuf.
+* platform - Handles some platform specific concepts, like a Window.
+* graphics - Graphics layer with e.g. vertex buffers and shader programs. Backends for OpenGL, Vulkan etc.
+* particle - A system for simulating particles. Use by comp_particle.cpp
```
Comment (britzl) [inline engine/docs/CODE_STRUCTURE.md:34]: Use by -> Used by

File: engine/docs/CODE_STRUCTURE.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+* lua - The regular Lua 5.1 implementation.
+* hid - Human Interface Device library that handles gamepads, mouse, keyboards and touch input.
+* input - A higher level input system that maps HID events into input actions
+* physics - A single frontend for our two physics engines Box 2D and Bullet 3D. Includes the source of Box2D
```
Comment (britzl) [inline engine/docs/CODE_STRUCTURE.md:38]: Users will ask where the source of Bullet 3D resides. We should mention that it is in packages.

File: engine/docs/CODE_STRUCTURE.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+* render - High level rendering concepts like materials, constants, cameras and render command buffer.
+* rig - Our model+skeleton animation system.
+* gameobject - Collections, game objects and component base functionality
+* gui - Our gui system. Used by `comp_gui.cpp`
+* [sound](../sound/README.md) - Our software sound mixing library
+* liveupdate - High level functionality for downloadable content
+* crash - Callstack callbacks and writing functionality
+* gamesys - The main bulk of all resource types, component types and script modules.
+* tools - Utilities like the `gdc` and the editor `launcher`
+* record - Video (.avi) recording library
```
Comment (britzl) [inline engine/docs/CODE_STRUCTURE.md:51]: "Current video format is the open vp8 codec in the ivf container"

File: engine/docs/CODE_STRUCTURE.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+### Code conventions
+
+This document isn't about the code style specifically, but it might be good to mention a few things.
+We have a detailed [blog post](https://defold.com/2020/05/31/The-Defold-engine-code-style/) about our code style, but here we list the key points.
+
+In our code we use C++ very sparingly.
+We mainly use it for namespaces, and our own container types.
+
+* No STL - We do not use STL in our code (with one single exception, and we should remove that!).
```
Comment (britzl) [inline engine/docs/CODE_STRUCTURE.md:109]: What is the exception?

**Review by britzl — APPROVED** [review]

**Review by britzl — CHANGES_REQUESTED** [review]
Check the comments

File: engine/docs/CODE_STRUCTURE.md (hunk 4 above)
Comment (JCash) [reply engine/docs/CODE_STRUCTURE.md:109]: That was intentionally left out to not encourage more usage of it.
But I'll just remove it :)

**Review by britzl — DISMISSED** [review]

**Pull request merged**